### PR TITLE
ESS_GUI 1381 slicer

### PR DESCRIPTION
**Pull request by rozyczko**
Revamped batch slicer from version 4.x

Context menu option for slicer edit shows up always for 2D plots.
Widget allows creating and/or deleting slicers (set to None).
Batch slicing and relevant options in separate tab.

### CODE REVIEW

**Comment by butlerpd** [comment]
5.0 ready for testing on Win

File: src/sas/qtgui/Plotting/SlicerParameters.py
Comment (krzywon) [inline src/sas/qtgui/Plotting/SlicerParameters.py:319]: This sends data to whichever perspective is currently active.

File: src/sas/qtgui/Plotting/SlicerParameters.py
Comment (krzywon) [inline src/sas/qtgui/Plotting/SlicerParameters.py:293]: Potential improvement (not necessarily here): Allow data to be saved as NXcanSAS or cansas XML.

**Review by krzywon — CHANGES_REQUESTED** [review]
Issues that need fixing: setting the slicer to None and not ensuring the fitting perspective is currently active.

More new unit tests would be helpful.

The ability to apply the slicer to any loaded 2D data (not just plotted data) and the ability to save the output in either cansas format would be helpful features, but neither is a showstopper.

**Comment by krzywon** [comment]
One last item - the save location is not being honored for me. Everything gets saved in the same directory as run.py.

File: src/sas/qtgui/MainWindow/media/graph_help.rst
Comment (rozyczko) [reply src/sas/qtgui/MainWindow/media/graph_help.rst:316]: ahem... it actualy was... I just copy/pasted 4.x docs changing some content to fit the new widget... sorry

File: src/sas/qtgui/Plotting/SlicerParameters.py
Comment (rozyczko) [reply src/sas/qtgui/Plotting/SlicerParameters.py:309]: Good catch. this worked for me when tested but apparently new stuff added later broke this.

File: src/sas/qtgui/Plotting/SlicerParameters.py
Comment (rozyczko) [reply src/sas/qtgui/Plotting/SlicerParameters.py:319]: Fitting perspective - best perspective.
Also, well spotted.

**Comment by wpotrzebowski** [comment]
ready for testing on Linux

File: src/sas/qtgui/Plotting/SlicerParameters.py
Comment (rozyczko) [reply src/sas/qtgui/Plotting/SlicerParameters.py:297]: Considering we give this option to all other 1D plots, this needs to be implemented here as well.

**Comment by butlerpd** [comment]
ready for testing on Win

File: src/sas/qtgui/Plotting/SlicerParameters.py
Comment (krzywon) [inline src/sas/qtgui/Plotting/SlicerParameters.py:312]: New DataExplorer objects are not appended with '[n]' so Data1D objects can end up with the same display name (same slicer, same data file, switching back and forth between slicers).

**Review by krzywon — CHANGES_REQUESTED** [review]
This is very close to ready, but there are a few additional issues I've found that I'd like to see addressed.

File: src/sas/qtgui/Plotting/SlicerParameters.py
Comment (rozyczko) [reply src/sas/qtgui/Plotting/SlicerParameters.py:312]: Yes, I noticed that just recently - will fix it

File: src/sas/qtgui/Plotting/SlicerParameters.py
Comment (rozyczko) [reply src/sas/qtgui/Plotting/SlicerParameters.py:198]: Yes, model param change should not trigger file resave. File resave should be done only on Apply. Fixed.
Also, on param change changes params for all selected plots now, as expected

File: src/sas/qtgui/Plotting/SlicerParameters.py
Comment (rozyczko) [reply src/sas/qtgui/Plotting/SlicerParameters.py:312]: Absolutely.
Fixed.

File: src/sas/qtgui/Plotting/SlicerParameters.py
Comment (rozyczko) [reply src/sas/qtgui/Plotting/SlicerParameters.py:291]: This is fixed now.

File: src/sas/qtgui/Plotting/SlicerParameters.py
Comment (rozyczko) [reply src/sas/qtgui/Plotting/SlicerParameters.py:110]: I'd prefer to do that as a separate task, if possible.

File: src/sas/qtgui/Plotting/SlicerParameters.py
Comment (krzywon) [reply src/sas/qtgui/Plotting/SlicerParameters.py:341]: I'm still seeing this. I applied a sector average to 3 plots and then switched to an annular average. When I clicked apply for the annular average, a second set of sector plots were created that were sent to the same batch fit as the annular averages.

BatchSlicer_GeneratingOldAndNew

File: src/sas/qtgui/Plotting/SlicerParameters.py
Comment (krzywon) [reply src/sas/qtgui/Plotting/SlicerParameters.py]: I'm seeing some coordination issues between the plot and slicer parameters window:

When you drag the slicer on the plot, the slicer parameters window is updated to match the plot values, but those values aren't then applied to other slicers. Only changes made directly in the slicer parameters window trigger changes to the selected plots. The 'Apply' button brings all plots back to the same state.

BatchSlicer_SlicerPlotMoved_NotCascading

Comment (krzywon) [reply src/sas/qtgui/Plotting/SlicerParameters.py]: Changing the slicer type on the plot with the slicer parameters window open updates the parameter list, but does not update the slicer type combo box. Clicking 'Apply' or changing a value in the parameter window changes the slicer to whichever one is selected in the combo box.

BatchSlicer_SlicerTypeNotUpdated

**Comment by murphyryanp** [comment]
ready for testing on Win

**Review by murphyryanp — APPROVED** [review]
Working great for me.

**Comment by butlerpd** [comment]
Overall I like it! will be much slicker than 4.x I think.  However I found 3 relatively minor problems.  Two should probably be fixed as part of this PR the third should be dealt with separately and needs consultation on what the desired behavior should be.

**Should Fix**
* clicking on "toggle log/linear" in the 2D plot context menu creates a circular average.  No idea how those got linked? Testing PR build 66 (invariant stuff) this does not happen, so assume that the cross talk is being introduced in this PR?
* This is a use case for issue #1566.   For example:
  * Plot a 2D data set.  Open a slicer. do whatever manipulations are desired.
  * At this point either delete all plots or not.  However, unselect the current data and select another 2D data set.
  * Now send this to data to plot (create new graph).  Not only does one have the new 2D data plot but **also** the slicers from the last data (which are all secretly stored and selected under the first 2D data).  This is most confusing as it appears as if the second go automatically selected and plotted the last slicer.  One has to look closely to realize that in fact it is a slicer on the **PREVIOUS** data.  Can we ensure that the daughter data/plots created are **not** selected when they are created? It should be up to the user to select them for plotting at some point **IF** desired.

**To Discuss**
The Batch function seems to be hidden unless you know it is there -- perhaps such a feature should not be used without knowing enough that one also knows it is there? Currently one has to click on `Edit Slicer Parameters` in the context menu and then click over to the Batch tab. It is not clear what the best behavior would be.  Perhaps renaming the menu item to `Slicers` and remove the other menu items?
  * The advantages would be:
     * Much simplified context menu
     * More obvious accessibility to Batch
     * Generally cleaner? 
  * The Disadvantage is that it would break from 4.x significantly and old users may get upset 
.

**Comment by rozyczko** [comment]
> * clicking on "toggle log/linear" in the 2D plot context menu creates a circular average.  No idea how those got linked? Testing PR build 66 (invariant stuff) this does not happen, so assume that the cross talk is being introduced in this PR?

Yes, the circ average plot shows on any replot function but only if it has been previously generated. Pretty sloppy, I'd say. Needs fixing ideed.

> * This is a use case for issue #1566.   For example:
>   
>   * Plot a 2D data set.  Open a slicer. do whatever manipulations are desired.
>   ...

This is similar in scope to what Jeff noticed in a comment above. Fixing now.


> **To Discuss**
> The Batch function seems to be hidden unless you know it is there -- perhaps such a feature should not be used without knowing enough that one also knows it is there? Currently one has to click on `Edit Slicer Parameters` in the context menu and then click over to the Batch tab. It is not clear what the best behavior would be. Perhaps renaming the menu item to `Slicers` and remove the other menu items?
> 
> * The advantages would be:
>   
>   * Much simplified context menu
>   * More obvious accessibility to Batch
>   * Generally cleaner?
> * The Disadvantage is that it would break from 4.x significantly and old users may get upset

I was thinking about use scenarios for this functionality and decided to leave it as it is now. The basic functionality of creating a simple slicer should be easily available, and we can do it now with context menu->slicer + manual dragging. This also maintains "backward functionality" as you said in the last point.
However, having redundancy in the context menu is not optimal (although it already exists to some extent with explicit plot options vs. mpl toolbar options). Removing the extra 4 slicer entries would simplify the menu.
It all really depends whether the shortcuts to simple slicing are being used heavily by users. If not, we can play the "new version gives you new access methods" card and redesign access to the slicers.

File: src/sas/qtgui/Plotting/SlicerParameters.py
Comment (rozyczko) [reply src/sas/qtgui/Plotting/SlicerParameters.py]: 1. Update the single slicer in real time: this is how I designed it - parameter changes affect the current plot and all plots get updated on Apply. In 4.x, after you modify parameters you need to click Apply to see them on all plots.
2. Combo box update will be addressed.Well spotted.

File: src/sas/qtgui/Plotting/SlicerParameters.py
Comment (rozyczko) [reply src/sas/qtgui/Plotting/SlicerParameters.py:341]: This is fixed now, attempt #2

**Comment by katieweigandt** [comment]
I'm playing with last night's build.  It's pretty awesome and mostly very intuitive.  The first time through I missed the define save path, but I used it on the 2nd pass and it worked great.  One tiny request.  I was batch slicing 60 files  and needed to click to include each of them. Could you add a select all option?

**Comment by butlerpd** [comment]
> > * This is a use case for issue #1566.   For example:
> >   
> >   * Plot a 2D data set.  Open a slicer. do whatever manipulations are desired.
> >     ...
> 
> @butlerpd , @krzywon :
> OK, now I remember how I designed this (sorry, it's been a while). This all stems from the requirement of having a non-modal dialog, since with modal dialogs things are MUCH more obvious and explicit.
> The question is - which plot is the **main** plot for deciding which changes go where?
> 
> * I open the widget from plot1 and then select plot1, plot2 and plot3 for batch slicing
> * I apply the same slicer to all three
> * I change the slicer in plot2 (using context menu in the plot). Should that change the widget content? What happens when I click Apply? Should the new slicer from plot2 propagate to plot1 and plot3 or should it revert to what plot1 originally had?
> * What if I change plot3 as well? Which slicer/parameters update what plots on Apply?
> 
> I chose the more intuitive (to me) option of having plot1 be the `master`, since I opened the slicer widget from it.
> This means every change to plot1 will propagate to all other plots on `Apply`. Same with parameter change.
> 
> Does this make sense? Have I missed a use case or workflow?
> 
Hummm... I see. This is something completely different from what I was discussing and something I did not even discover in my testing.  Clearly my testing was not very thorough :-(   That said I think my instinct would be the opposite: the apply should **only** apply to the slicer plot that is checked in the batch tab. I would presume that by default the batch tab only has the current data set selected (i.e. the data set rom which I launched the `Edit Slicer Parameters` dialog.

Perhaps we can have a quick meeting but am thinking maybe we merge this and start a new issue PR to fix anything else? Including Katie's request for a "select/unselect" all button?

**Comment by butlerpd** [comment]
also @rozyczko it looks like there is a conflict now with the dataexplorer.py that will need to be resolved?

**Review by krzywon — APPROVED** [review]

**Comment by krzywon** [comment]
I just tested the latest and agree with Paul, this can be merged once the merge conflicts are resolved.

**Pull request merged**
